// @twin.org/identity-models/dist/cjs/index.cjs
'use strict';

var core = require('@twin.org/core');
var standardsW3cDid = require('@twin.org/standards-w3c-did');
var web = require('@twin.org/web');

// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
/**
 * Factory for creating identity connectors.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const IdentityConnectorFactory = core.Factory.createFactory("identity-connector");

// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
/**
 * Factory for creating identity profile connectors.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const IdentityProfileConnectorFactory = core.Factory.createFactory("identity-profile-connector");

// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
/**
 * Factory for creating identity resolver connectors.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const IdentityResolverConnectorFactory = core.Factory.createFactory("identity-resolver-connector");

// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
/**
 * Helper methods for documents.
 */
class DocumentHelper {
    /**
     * Runtime name for the class.
     */
    static CLASS_NAME = "DocumentHelper";
    /**
     * Parse the document id into its parts.
     * @param documentId The full document id.
     * @returns The parsed document id.
     */
    static parseId(documentId) {
        if (!core.Is.stringValue(documentId)) {
            return {
                id: "",
                fragment: ""
            };
        }
        const fragmentIndex = documentId.indexOf("#");
        return {
            id: fragmentIndex === -1 ? documentId : documentId.slice(0, fragmentIndex),
            fragment: fragmentIndex === -1 ? undefined : documentId.slice(fragmentIndex + 1)
        };
    }
    /**
     * Join the document id parts.
     * @param documentId The document id.
     * @param fragment The fragment part for the identifier.
     * @returns The full id.
     */
    static joinId(documentId, fragment) {
        if (!core.Is.stringValue(documentId)) {
            return "";
        }
        let fullId = documentId;
        if (core.Is.stringValue(fragment)) {
            if (fragment.startsWith(documentId)) {
                fragment = fragment.slice(documentId.length);
            }
            if (fragment.startsWith("#")) {
                fullId += fragment;
            }
            else {
                fullId += `#${fragment}`;
            }
        }
        return fullId;
    }
    /**
     * Get a verification method from a DID document.
     * @param didDocument The DID Document to get the method from.
     * @param methodName The name of the method to get the JWK from.
     * @param methodType The type of the method, defaults to verificationMethod.
     * @returns The verification method if found.
     * @throws Error if the method is not found.
     */
    static getVerificationMethod(didDocument, methodName, methodType) {
        const verificationMethod = didDocument[methodType ?? standardsW3cDid.DidVerificationMethodType.VerificationMethod]?.find(vm => core.Is.object(vm) && vm.id === methodName);
        if (core.Is.object(verificationMethod)) {
            return verificationMethod;
        }
        throw new core.GeneralError(DocumentHelper.CLASS_NAME, "verificationMethodNotFound", {
            methodName,
            methodType
        });
    }
    /**
     * Gets a JWK from a DID document verification method.
     * @param didDocument The DID Document to get the method from.
     * @param methodName The name of the method to get the JWK from.
     * @param methodType The type of the method, defaults to verificationMethod.
     * @returns The JWK if found.
     * @throws Error if the method is not found.
     */
    static getJwk(didDocument, methodName, methodType) {
        const verificationMethod = DocumentHelper.getVerificationMethod(didDocument, methodName, methodType);
        if (core.Is.object(verificationMethod) && core.Is.object(verificationMethod.publicKeyJwk)) {
            return verificationMethod.publicKeyJwk;
        }
        throw new core.GeneralError(DocumentHelper.CLASS_NAME, "verificationMethodJwkNotFound", {
            methodName,
            methodType
        });
    }
}

// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
/**
 * Helper methods for verification.
 */
class VerificationHelper {
    /**
     * Runtime name for the class.
     */
    static CLASS_NAME = "VerificationHelper";
    /**
     * Verified the JWT.
     * @param resolver The resolver to use for finding the document.
     * @param jwt The token to verify.
     * @returns The decoded payload.
     */
    static async verifyJwt(resolver, jwt) {
        core.Guards.object(VerificationHelper.CLASS_NAME, "resolver", resolver);
        core.Guards.string(VerificationHelper.CLASS_NAME, "jwt", jwt);
        const jwtDecoded = await web.Jwt.decode(jwt);
        const jwtHeader = jwtDecoded.header;
        const jwtPayload = jwtDecoded.payload;
        const jwtSignature = jwtDecoded.signature;
        if (!core.Is.object(jwtHeader) || !core.Is.object(jwtPayload) || !core.Is.uint8Array(jwtSignature)) {
            throw new core.GeneralError(VerificationHelper.CLASS_NAME, "jwtDecodeFailed");
        }
        const iss = jwtHeader?.iss;
        const kid = jwtHeader?.kid;
        core.Guards.stringValue(VerificationHelper.CLASS_NAME, "iss", iss);
        core.Guards.stringValue(VerificationHelper.CLASS_NAME, "kid", kid);
        const didDocument = await resolver.identityResolve(iss);
        const jwk = DocumentHelper.getJwk(didDocument, kid);
        const publicKey = await web.Jwk.toCryptoKey(jwk);
        return web.Jwt.verify(jwt, publicKey);
    }
    /**
     * Verified the proof for the document e.g. verifiable credential.
     * @param resolver The resolver to use for finding the document.
     * @param secureDocument The secure document to verify.
     * @returns True if the verification is successful.
     */
    static async verifyProof(resolver, secureDocument) {
        core.Guards.object(VerificationHelper.CLASS_NAME, "resolver", resolver);
        core.Guards.object(VerificationHelper.CLASS_NAME, "secureDocument", secureDocument);
        core.Guards.object(VerificationHelper.CLASS_NAME, "secureDocument.proof", secureDocument.proof);
        const proofList = core.Is.array(secureDocument.proof)
            ? secureDocument.proof
            : [secureDocument.proof];
        const documentCache = {};
        for (const proof of proofList) {
            if (!core.Is.stringValue(proof?.verificationMethod)) {
                throw new core.GeneralError(VerificationHelper.CLASS_NAME, "proofMissingVerificationMethod");
            }
            const proofVerificationMethod = DocumentHelper.parseId(proof.verificationMethod);
            if (!core.Is.stringValue(proofVerificationMethod.fragment)) {
                throw new core.GeneralError(VerificationHelper.CLASS_NAME, "proofMissingVerificationMethod");
            }
            let document;
            if (documentCache[proofVerificationMethod.id]) {
                document = documentCache[proofVerificationMethod.id];
            }
            else {
                document = await resolver.identityResolve(proofVerificationMethod.id);
                documentCache[proofVerificationMethod.id] = document;
            }
            const verificationJwk = await DocumentHelper.getJwk(document, proofVerificationMethod.id);
            const verified = standardsW3cDid.ProofHelper.verifyProof(secureDocument, proof, verificationJwk);
            if (!verified) {
                return false;
            }
        }
        return true;
    }
}

exports.DocumentHelper = DocumentHelper;
exports.IdentityConnectorFactory = IdentityConnectorFactory;
exports.IdentityProfileConnectorFactory = IdentityProfileConnectorFactory;
exports.IdentityResolverConnectorFactory = IdentityResolverConnectorFactory;
exports.VerificationHelper = VerificationHelper;
