// @iota/identity-wasm/web/controller.js
/**
 * Permissions of a {@link DelegationToken}.
 */
export var DelegatePermissions;
(function (DelegatePermissions) {
    /** No permissions. */
    DelegatePermissions[DelegatePermissions["None"] = 0] = "None";
    /** Delegate can create new proposals. */
    DelegatePermissions[DelegatePermissions["CreateProposal"] = 1] = "CreateProposal";
    /** Delegate can approve existing proposals. */
    DelegatePermissions[DelegatePermissions["ApproveProposal"] = 2] = "ApproveProposal";
    /** Delegate can execute proposals. */
    DelegatePermissions[DelegatePermissions["ExecuteProposal"] = 4] = "ExecuteProposal";
    /** Delegate can delete proposals. */
    DelegatePermissions[DelegatePermissions["DeleteProposal"] = 8] = "DeleteProposal";
    /** Delegate can remove its controller's approval. */
    DelegatePermissions[DelegatePermissions["RemoveApproval"] = 16] = "RemoveApproval";
    /** All permissions. */
    DelegatePermissions[DelegatePermissions["All"] = 4294967295] = "All";
})(DelegatePermissions || (DelegatePermissions = {}));
