// @twin.org/logging-models/dist/cjs/index.cjs
'use strict';

var core = require('@twin.org/core');

// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
/**
 * Factory for creating logging connectors.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const LoggingConnectorFactory = core.Factory.createFactory("logging");

// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
/**
 * Class for performing logging operations on multiple connectors.
 */
class MultiLoggingConnector {
    /**
     * The namespace for the logging connector.
     */
    static NAMESPACE = "multi";
    /**
     * Runtime name for the class.
     */
    CLASS_NAME = "MultiLoggingConnector";
    /**
     * The connectors to send the log entries to.
     * @internal
     */
    _loggingConnectors;
    /**
     * The log levels to display, will default to all.
     * @internal
     */
    _levels;
    /**
     * Create a new instance of MultiLoggingConnector.
     * @param options The options for the connector.
     */
    constructor(options) {
        core.Guards.object(this.CLASS_NAME, "options", options);
        core.Guards.arrayValue(this.CLASS_NAME, "options.loggingConnectorTypes", options.loggingConnectorTypes);
        this._levels = options?.config?.levels ?? ["debug", "info", "warn", "error", "trace"];
        this._loggingConnectors = options.loggingConnectorTypes.map(t => LoggingConnectorFactory.get(t));
    }
    /**
     * Log an entry to the connector.
     * @param logEntry The entry to log.
     * @returns Nothing.
     */
    async log(logEntry) {
        core.Guards.object(this.CLASS_NAME, "logEntry", logEntry);
        if (this._levels.includes(logEntry.level)) {
            logEntry.ts ??= Date.now();
            await Promise.allSettled(this._loggingConnectors.map(async (loggingConnector) => loggingConnector.log(logEntry)));
        }
    }
    /**
     * Query the log entries.
     * @param conditions The conditions to match for the entities.
     * @param sortProperties The optional sort order.
     * @param properties The optional keys to return, defaults to all.
     * @param cursor The cursor to request the next page of entities.
     * @param pageSize The maximum number of entities in a page.
     * @returns All the entities for the storage matching the conditions,
     * and a cursor which can be used to request more entities.
     */
    async query(conditions, sortProperties, properties, cursor, pageSize) {
        // See if we can find a connector that supports querying.
        // If it throws anything other than not implemented, we should throw it.
        for (const loggingConnector of this._loggingConnectors) {
            // eslint-disable-next-line @typescript-eslint/unbound-method
            if (core.Is.function(loggingConnector.query)) {
                return loggingConnector.query(conditions, sortProperties, properties, cursor, pageSize);
            }
        }
        return {
            entities: []
        };
    }
}

/**
 * Class for performing logging operations to nowhere.
 */
class SilentLoggingConnector {
    /**
     * The namespace for the logging connector.
     */
    static NAMESPACE = "silent";
    /**
     * Runtime name for the class.
     */
    CLASS_NAME = "SilentLoggingConnector";
    /**
     * Log an entry to the connector.
     * @param logEntry The entry to log.
     * @returns Nothing.
     */
    async log(logEntry) { }
    /**
     * Query the log entries.
     * @param conditions The conditions to match for the entities.
     * @param sortProperties The optional sort order.
     * @param properties The optional keys to return, defaults to all.
     * @param cursor The cursor to request the next page of entities.
     * @param pageSize The maximum number of entities in a page.
     * @returns All the entities for the storage matching the conditions,
     * and a cursor which can be used to request more entities.
     */
    async query(conditions, sortProperties, properties, cursor, pageSize) {
        return {
            entities: []
        };
    }
}

// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
/**
 * Helper class for log entry operations.
 */
class LogEntryHelper {
    /**
     * Translate the log entry.
     * @param logEntry The log entry.
     * @returns The translated log entry if a translation can be found.
     */
    static translate(logEntry) {
        if (core.Is.stringValue(logEntry.level) && core.Is.stringValue(logEntry.source)) {
            const sourceMessage = `${logEntry.level}.${core.StringHelper.camelCase(logEntry.source)}.${logEntry.message}`;
            if (core.I18n.hasMessage(sourceMessage)) {
                return core.I18n.formatMessage(sourceMessage, logEntry.data);
            }
        }
        if (core.Is.stringValue(logEntry.source)) {
            const sourceMessage = `${core.StringHelper.camelCase(logEntry.source)}.${logEntry.message}`;
            if (core.I18n.hasMessage(sourceMessage)) {
                return core.I18n.formatMessage(sourceMessage, logEntry.data);
            }
        }
        if (core.I18n.hasMessage(logEntry.message)) {
            return core.I18n.formatMessage(logEntry.message, logEntry.data);
        }
    }
}

exports.LogEntryHelper = LogEntryHelper;
exports.LoggingConnectorFactory = LoggingConnectorFactory;
exports.MultiLoggingConnector = MultiLoggingConnector;
exports.SilentLoggingConnector = SilentLoggingConnector;
