// @iota/identity-wasm/node/key_id_storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyIdMemStore = void 0;
const base64_arraybuffer_1 = require("base64-arraybuffer");
class KeyIdMemStore {
    _keyIds;
    constructor() {
        this._keyIds = new Map();
    }
    async insertKeyId(methodDigest, keyId) {
        let methodDigestAsString = methodDigestToString(methodDigest);
        let value = this._keyIds.get(methodDigestAsString);
        if (value !== undefined) {
            throw new Error("KeyId already exists");
        }
        this._keyIds.set(methodDigestAsString, keyId);
    }
    async getKeyId(methodDigest) {
        let methodDigestAsString = methodDigestToString(methodDigest);
        let value = this._keyIds.get(methodDigestAsString);
        if (value == undefined) {
            throw new Error("KeyId not found");
        }
        return value;
    }
    async deleteKeyId(methodDigest) {
        let methodDigestAsString = methodDigestToString(methodDigest);
        let success = this._keyIds.delete(methodDigestAsString);
        if (success) {
            return;
        }
        else {
            throw new Error("KeyId not found!");
        }
    }
    count() {
        return this._keyIds.size;
    }
}
exports.KeyIdMemStore = KeyIdMemStore;
/**
 * Converts a `MethodDigest` to a base64 encoded string.
 */
function methodDigestToString(methodDigest) {
    let arrayBuffer = methodDigest.pack().buffer;
    return (0, base64_arraybuffer_1.encode)(arrayBuffer);
}
